source/mir/random/package.d

```diff
@@ -317,16 +317,46 @@ Returns:
 T randIndex(T, G)(ref G gen, T m)
     if(isSaturatedRandomEngine!G && isUnsigned!T)
 {
-    assert(m, "m must be positive");
-    T ret = void;
-    T val = void;
-    do
+    alias R = EngineReturnType!G;
+    version (D_X32)
+        enum bool versionD_X32 = true;//On 64-bit machine but using 32-bit pointers.
+    else
+        enum bool versionD_X32 = false;
+    static if (R.sizeof >= T.sizeof * 2 && (size_t.sizeof >= R.sizeof || versionD_X32))
```

Comment:
- The original algorithm generates the same amount of bits as the result has. It do not need to generate more. - So, it should be just `gen.rand!T`  instead of `gen() >>> rshift` : `rand` will chose lowest or highest bits depends on the engine.
- We only need to check that `multiresult` has at least twice more bits than R.

The problem for this optimisation is in practice `T` is `size_t`. In other hand `x86` and `x86_64` generates extended multiplication results stored in two registers: 32x32->64 and 64x64->128 (x86_64).

- ubyte and ushort T: `randombits * m;` (the result will be 32 bit because of language semantics)
- uint T needs `randombits * cast(ulong) m;` and compiler will do its best job for all targets.
- ulong T needs to use LLVM IR if no `ucent` or `randombits * cast(ucent) m;` if ucent
- ucent T need to use LLVM IR

LLVM IR
 - https://wiki.dlang.org/LDC_inline_IR
 - https://llvm.org/docs/LangRef.html

Reply:
- Please add `@safe` attribute for unittests.
- Please addd CTFE unittests. LLVM IR can no be executed compile time, so the old code should be used in CTFE (always for simplicity).

Reply:
`= void` probably unsafe, but I am not sure